libyara/ahocorasick.c

```diff
@@ -209,19 +363,524 @@ static int _yr_ac_state_destroy(YR_AC_STATE* state)
+      }
+    }  // else if (input_state->type == YR_ATOM_TYPE_CLASS)
+  }    // else if (next_state->type == YR_ATOM_TYPE_CLASS)
+
+  return true;
+}
+
+////////////////////////////////////////////////////////////////////////////////
 //
-// This function must be called after all the strings have been added to the
-// automaton with yr_ac_add_string.
+// In case of collision, some symbols have to be excluded from the state,
+// and the copy of the state is created
 //
-static int _yr_ac_create_failure_links(YR_AC_AUTOMATON* automaton)
+static bool _yr_ac_exclude_from_state(
+    YR_AC_STATE* input_state,
+    YR_AC_STATE* dfa_state1,
+    YR_AC_STATE* dfa_state2,
+    YR_ARENA* arena)
+{
+  YR_AC_STATE* next1 = NULL;
+  YR_AC_STATE* next2 = NULL;
+  YR_AC_STATE* temp_state;
+  YR_AC_STATE* prev_state = NULL;
+
+  next1 = _yr_ac_next_state_typed(
+      input_state->parent,
+      dfa_state1->input,
+      dfa_state1->type,
+      dfa_state1->bitmap);
+  next2 = _yr_ac_next_state_typed(
+      input_state->parent,
+      dfa_state2->input,
+      dfa_state2->type,
+      dfa_state2->bitmap);
+
+  if ((next1 != NULL) && (next2 != NULL))
+  {
+    // Both version of the input_state already exist in the AC
+    _yr_ac_copy_path(input_state, input_state->parent, dfa_state1, arena);
+    _yr_ac_copy_path(input_state, input_state->parent, dfa_state2, arena);
+    temp_state = input_state->parent->first_child;
+
+    if (temp_state == input_state)
+    {
+      input_state->parent->first_child = temp_state->siblings;
+      _yr_ac_state_destroy_updated(input_state);
+      input_state->type = YR_ATOM_TYPE_REMOVE;
+      return true;
+    }
+
+    while (temp_state != NULL && temp_state != input_state)
+    {
+      prev_state = temp_state;
+      temp_state = temp_state->siblings;
+    }
+
+    if (prev_state != NULL)
+      prev_state->siblings = temp_state->siblings;
+
+    _yr_ac_state_destroy_updated(input_state);
+    input_state->type = YR_ATOM_TYPE_REMOVE;
+    return true;
+  }
+  else if (next2 != NULL)
+  {
+    // The updated state aleady exists in the AC
+    memcpy(input_state, dfa_state1, sizeof(YR_AC_STATE));
+    _yr_ac_copy_path(input_state, input_state->parent, dfa_state2, arena);
+  }
+  else
+  {
+    // Neither version of the input_state already exist in the AC or there is
+    // only literal state
+    memcpy(input_state, dfa_state2, sizeof(YR_AC_STATE));
+    _yr_ac_copy_path(input_state, input_state->parent, dfa_state1, arena);
+  }
+
+  return false;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+// Creates deterministic AC for failure function
+//
+static bool dfa_subtree(
+    YR_AC_STATE* state,
+    YR_AC_STATE* input_state,
+    YR_ARENA* arena)
+{
+  bool pop = true;
+
+  if (input_state->type == YR_ATOM_TYPE_LITERAL)
+    return pop;
+
+  int literal = _yr_ac_class_is_literal(input_state->bitmap);
+  if (literal >= 0)
+  {
+    input_state->input = literal;
```

Comment:
`input_state` is modified here, that's a side effect of `dfa_subtree` that should avoided. Why is this required here? Is it possible to convert all the states of type `YR_ATOM_TYPE_CLASS` that can be converted to `YR_ATOM_TYPE_LITERAL` in a single place?

Reply:
The change to the literal state is needed because it influences the algorithm. I tried to revise the algorithm several times, but I am mainly facing problems with match structure. It is sensitive to changes and copies, and many changes in the algorithm influence the order of matches, for example.